### PR TITLE
Add numerical stability to SoftmaxGrad test inputs.

### PR DESCRIPTION
**Pull request by codemzs**
**SoftmaxGrad** tests that verify gradients computed from CPU and GPU kernels match have been observed to fail sporadically on Linux and Windows. The issue seems to arise due to the absence of numerical stability in test input `Y` that PyTorch [softmax_grad test](https://github.com/pytorch/pytorch/blob/be757957bace28100e571ec7914765020be4a069/caffe2/python/operator_test/softmax_ops_test.py#L69) seems to add. In addition to this we must restrict the domain of input values from **[-10.0, 10.0)** to **[0, 1)** just like PyTorch [softmax_grad test](https://github.com/pytorch/pytorch/blob/be757957bace28100e571ec7914765020be4a069/caffe2/python/operator_test/softmax_ops_test.py#L69).

**Just by alone adding `1e-2` to input `Y` for numerical stability I was able to resolve below failures.** I have also verified both CPU and GPU implementation are correct. To further verify the correctness of this change I ran both the tests in a loop for 1000 with varying random seed and did not observe any failures on Linux. This change also re-enables SoftmaxGrad_LargeTensor test.

**CudaKernelTest.SoftmaxGrad_SmallTensor**
```
Linux:
2020-05-04T22:35:08.3699287Z 1: expected 2.33688 (40158f69), got 2.2905 (4012979b), diff: 0.0463748, tol=0.0117844. 2939 of 262144 differ
ORT test random seed: 3190010571

Windows:
expected 5.29682 (40a97f97), got 5.42497 (40ad995c), diff: 0.128146, tol=0.0265841. 3518 of 262144 differ

ORT test random seed: 1224836532
```
**CudaKernelTest.SoftmaxGrad_LargeTensor**
```
Linux:
random seed: 552621640
```

### CODE REVIEW

File: orttraining/orttraining/test/training_ops/cuda/softmax_test.cc [hunk 1]
```diff
@@ -55,25 +55,28 @@ static void TestSoftmaxGrad(const std::vector<int64_t>& dY_dims,
   test.CompareWithCPU(kCudaExecutionProvider, per_sample_tolerance, relative_per_sample_tolerance);
 }
 
+// TODO fix flaky test
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/training_ops/cuda/softmax_test.cc:58]: If this comment is enough to don't forget about this test while keeping a better coverage, I am ok with it.

Comment (edgchen1) [reply orttraining/orttraining/test/training_ops/cuda/softmax_test.cc:58]: if we are increasing tolerances to mitigate this, the original tolerances that these random seed values cause failures for should be mentioned

File: orttraining/orttraining/test/training_ops/cuda/softmax_test.cc [hunk 2]
```diff
@@ -55,25 +55,28 @@ static void TestSoftmaxGrad(const std::vector<int64_t>& dY_dims,
...
+// failing on Linux random seed: 3190010571
+// failing on Windows random seed: 1224836532
 TEST(CudaKernelTest, SoftmaxGrad_SmallTensor) {
   std::vector<int64_t> dY_dims{8, 2, 128, 128};
   std::vector<int64_t> Y_dims{8, 2, 128, 128};
   std::vector<int64_t> dX_dims{8, 2, 128, 128};
 
-  const double per_sample_tolerance = 1e-4;
-  const double relative_per_sample_tolerance = 5e-3;
+  const double per_sample_tolerance = 1e-1;
```
Comment (edgchen1) [inline orttraining/orttraining/test/training_ops/cuda/softmax_test.cc:66]: 1e-4 to 1e-1 is a relatively large increase. perhaps just use fixed data to stabilize this test temporarily.

Comment (codemzs) [reply orttraining/orttraining/test/training_ops/cuda/softmax_test.cc:66]: Probably. I’m investigating the instability a little more now.

File: orttraining/orttraining/test/training_ops/cuda/softmax_test.cc (hunk 1 above)
Comment (codemzs) [reply orttraining/orttraining/test/training_ops/cuda/softmax_test.cc]: I'm not increasing the tolerance anymore.

File: orttraining/orttraining/test/training_ops/cuda/softmax_test.cc (hunk 2 above)
Comment (codemzs) [reply orttraining/orttraining/test/training_ops/cuda/softmax_test.cc]: The issue seems to be because of numerical instability in test data.

File: orttraining/orttraining/test/training_ops/cuda/softmax_test.cc [hunk 3]
```diff
@@ -44,8 +44,13 @@ static void TestSoftmaxGrad(const std::vector<int64_t>& dY_dims,
...
   // create rand inputs
   RandomValueGenerator random{};
-  std::vector<float> dY_data = random.Uniform<float>(dY_dims, -10.0f, 10.0f);
-  std::vector<float> Y_data = random.Uniform<float>(Y_dims, -10.0f, 10.0f);
+  std::vector<float> dY_data = random.Uniform<float>(dY_dims, 0.0f, 1.0f);
+  std::vector<float> Y_data = random.Uniform<float>(Y_dims, 0.0f, 1.0f);
+  // Add 1e-2 for numerical stability
+  for (int index = 0; index < Y_data.size(); index += 1) {
+    Y_data[index] += 1e-2f;
+  }
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/training_ops/cuda/softmax_test.cc:52]: Great investigation!!

FYI, these few line can also be written in
std::vector<float> Y_data = random.Uniform<float>(Y_dims, 0.02f, 1.02f);

Also, a even more realistic test data is to ensure Sum(Yi) = 1, where Yi is in [0, 1], since Y is the output of softmax. But this property of Y is not needed to ensure the numerical stabilyt, thus also skipped by pytorch test.

File: orttraining/orttraining/test/training_ops/cuda/softmax_test.cc [hunk 4]
```diff
@@ -65,9 +70,7 @@ TEST(CudaKernelTest, SoftmaxGrad_SmallTensor) {
   TestSoftmaxGrad(dY_dims, Y_dims, dX_dims, per_sample_tolerance, relative_per_sample_tolerance);
 }
 
-// TODO fix flaky test
-// failing random seed: 552621640
-TEST(CudaKernelTest, DISABLED_SoftmaxGrad_LargeTensor) {
+TEST(CudaKernelTest, SoftmaxGrad_LargeTensor) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/training_ops/cuda/softmax_test.cc:73]: With the fix, do we still need to relax the relative_per_sample_tolerance to make this test pass?

**Comment by Tixxx** [comment]
Thanks for the investigation. It’s a relief to know the implementations match. I guess it’s reasonable to add 1e-2 to input to prevent that probability from being ignored. Maybe we also need the stability term for other gradient tests that use 1e-1 threshold. We can take at look at that later.

File: orttraining/orttraining/test/training_ops/cuda/softmax_test.cc (hunk 3 above)
Comment (codemzs) [reply orttraining/orttraining/test/training_ops/cuda/softmax_test.cc:52]: Thanks!

File: orttraining/orttraining/test/training_ops/cuda/softmax_test.cc (hunk 4 above)
Comment (codemzs) [reply orttraining/orttraining/test/training_ops/cuda/softmax_test.cc:73]: I have changed relative_per_sample tolerance from 5e-3 to 1e-4 to keep it in parity with PyTorch [test](https://github.com/pytorch/pytorch/blob/ee23944f46a89208ccba26c8fa0c97fe585265ac/caffe2/python/serialized_test/serialized_test_util.py#L251), by doing so we have tightened the relative_per_sample_tolerance and tests still pass when I run them in a loop for 1000 times (small tensor) and 100 times (large tensor).

File: orttraining/orttraining/test/training_ops/cuda/softmax_test.cc [hunk 5]
```diff
@@ -44,8 +44,9 @@ static void TestSoftmaxGrad(const std::vector<int64_t>& dY_dims,
 
   // create rand inputs
   RandomValueGenerator random{};
-  std::vector<float> dY_data = random.Uniform<float>(dY_dims, -10.0f, 10.0f);
-  std::vector<float> Y_data = random.Uniform<float>(Y_dims, -10.0f, 10.0f);
+  std::vector<float> dY_data = random.Uniform<float>(dY_dims, 0.02f, 1.02f);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/training_ops/cuda/softmax_test.cc:47]: nit. the offset of 0.02 is only need for Y_data. 
Please also keep a comment explaining why we need this

Comment (codemzs) [reply orttraining/orttraining/test/training_ops/cuda/softmax_test.cc:47]: sorry, that was on accident. thanks.
